submissions/LuckyDnepr/OOP-Exercise/index.js

```diff
@@ -0,0 +1,144 @@
+"use strict";
+
+class Inhabitant {
+  constructor(species, name) {
+    this.species = species;
+    this.name = name;
+  }
+  getInfo() {
+    return `I'm ${this.species} and my name is ${this.name}.`;
+  }
+}
+
+class Saying {
+  constructor (person, vocabulary) {
+    return function (aventForAnswer) {
+      return person[vocabulary].hasOwnProperty(aventForAnswer)
+      ? person[vocabulary][aventForAnswer]
+      : "Nothing to say";
+    };
+  }
+}
+
+class Humans extends Inhabitant {
+  constructor(name, gender) {
+    super("Human", name);
+    this.gender = gender;
+    this.hands = 2;
+    this.legs = 2;
+  }
+  getInfo() {
+    return (
+      super.getInfo() +
+      ` I'm ${this.gender} with ${this.hands} hands and ${this.legs} legs.`
+    );
+  }
+}
+
+class FourLegged extends Inhabitant {
+  constructor(name, gender) {
+    super("Four-legged animal", name);
+    this.gender = gender;
+    this.paws = 4;
+  }
+  getInfo() {
+    return super.getInfo() + ` I'm ${this.gender} with ${this.paws} paws.`;
+  }
+}
+
+class CatlikeMutants extends Inhabitant {
+  constructor(gender, name) {
+    super("Mutant", name);
+    this.gender = gender;
+    this.hands = 2;
+    this.legs = 2;
+    //this.say = this.say.bind(new Cat());
+  }
+  getInfo() {
+    return (
+      super.getInfo() +
+      ` I'm ${this.gender} with ${this.hands} hands and ${this.legs} legs.`
+    );
+  }
+}
+
+class Man extends Humans {
+  constructor(name) {
+    super(name, "male");
+    this.vocabulary = {
+      hi: "Hey guys!",
+    };
+    this.say = new Saying(this, "vocabulary");
+  }
+  getInfo() {
+    return `${this.say("hi")} ` + super.getInfo();
+  }
+}
+
+class Woman extends Humans {
```

Comment:
Why men and women have their own classes while dogs and cats do not have gender-specific classes?

Reply:
In this case my reasoning was as follows: men and women can have specific methods in terms of social interaction in the community.
A dog and a cat most often do not have specific social differences in terms of gender.
Although, for certain tasks and more specific characteristics of the virtual world, I admit that the existence of classes "FemaleCat", "MaleDog", etc.

Reply:
> In this case my reasoning was as follows: men and women can have specific methods in terms of social interaction in the community.
What is the difference in these regards in current implementation?
Think of creating many women and men. And we also do not want a class per different values of the same property.

> A dog and a cat most often do not have specific social differences in terms of gender.
Oh, in general they do :)

Reply:
The separation of male and female was made to expand physical characteristics. For example, for the class "Woman" you can add the property women.breastSize = "2" ;-)

Reply:
This also would be a pre-mature optimization.
We need to stop somewhere within the given scope of task. Otherwise we end up having the class hiererachy representing the entire taxonomy of the real world.
On a real commercial project (e.g. a web-game) you will not build classes for all possible game objects that are not yet requested by the product team. Anyway they may come with something you didn't expect.

Keep class structure fitting existing requirements and yet scalable is what following SOLID and OOP principles help to achieve without excessive entities.

Reply:
I am not saying that current approach is illegal, but think of scalability and peer developers in a similar situation.

Reply:
Totally agree. I've had more than three choices, with different levels of class divisions. I settled on the minimum package, from my point of view. I understand that it is not optimal. My solution was focused on the task at hand.